### PR TITLE
Wallet NFT proposal

### PR DESCRIPTION
**Pull request by victormunoz**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-walletnft.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+---
+eip: <to be assigned>
+title: NFT with wallet
+description: NFT (ERC721) with only one token minted
+author: Victor Munoz (@victormunoz)
+discussions-to: https://ethereum-magicians.org/t/erc721-minting-only-one-token/8602/2
+status: Draft
+type: Standards
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:8]:
```suggestion
type: Standards Track
```

File: EIPS/eip-walletnft.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
+---
+eip: <to be assigned>
+title: NFT with wallet
+description: NFT (ERC721) with only one token minted
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:4]:
```suggestion
description: An ERC721-compatible single-token NFT
```

File: EIPS/eip-walletnft.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+author: Victor Munoz (@victormunoz)
+discussions-to: https://ethereum-magicians.org/t/erc721-minting-only-one-token/8602/2
+status: Draft
+type: Standards
+category (*only required for Standards Track): ERC
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:9]:
```suggestion
category: ERC
```

File: EIPS/eip-walletnft.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+created: 2022-03-25
+requires (*optional): 721
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:11]:
```suggestion
requires: 721
```

File: EIPS/eip-walletnft.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+---
+
+## Abstract
+NFT associated uniquely with a single contract address.
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:15]: This needs to be longer and more descriptive. See other Final EIPs for example abstracts.

File: EIPS/eip-walletnft.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+## Rationale
+The main motivation is to keep the contract compatible with current ERC721 platforms. That's why we maintain all the functions of the ERC721 standard even if they cannot be used (eg. mint) instead of just removing them.
+
+## Backwards Compatibility
+No backwards compatibility issues devised.
+
+## Reference Implementation
+Add the variable `_minted` in the contract:
+
+    bool private _minted;
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:38]:
```suggestion
``` solidity
    bool private _minted;
```a
```

(Remove the `a` afterwards. GitHub doesn't like closing code blocks)

File: EIPS/eip-walletnft.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+In the constructor, automint the first token and set the variable to true:
+
+    constructor(string memory name, string memory symbol, string memory base_uri) ERC721(name, symbol) {
+        baseUri = base_uri;
+        mint(msg.sender,0);
+        _minted = true;
+    }
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:46]:
```suggestion
``` solidity
    constructor(string memory name, string memory symbol, string memory base_uri) ERC721(name, symbol) {
        baseUri = base_uri;
        mint(msg.sender,0);
        _minted = true;
    }
```a
```

(Remove the `a` afterwards. GitHub doesn't like closing code blocks)

File: EIPS/eip-walletnft.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+Add additional functions to interact with the NFT properties (for instance, ERC20):
+
+    modifier onlyOwner() {
+        require(balanceOf(msg.sender) > 0, "Caller is not the owner of the NFT");
+        _;
+    }
+
+    function transferTokens(IERC20 token, address recipient, uint256 amount) public onlyOwner {
+        token.transfer(recipient, amount);
+    }
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:57]:
```suggestion
``` solidity
    modifier onlyOwner() {
        require(balanceOf(msg.sender) > 0, "Caller is not the owner of the NFT");
        _;
    }

    function transferTokens(IERC20 token, address recipient, uint256 amount) public onlyOwner {
        token.transfer(recipient, amount);
    }
```a
```
(Remove the `a` afterwards. GitHub doesn't like closing code blocks)

File: EIPS/eip-walletnft.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+## Specification
+The constructor should mint the unique token of the contract, and then the mint function should add a restriction to avoid further minting.
+
+Also, a `tokenTransfer` function should be added in order to allow the contract owner to transact with the ERC20 tokens owned by the contract/NFT itself.
+
+## Rationale
+The main motivation is to keep the contract compatible with current ERC721 platforms. That's why we maintain all the functions of the ERC721 standard even if they cannot be used (eg. mint) instead of just removing them.
+
+## Backwards Compatibility
+No backwards compatibility issues devised.
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:33]:
```suggestion
There are no backwards compatibility issues.
```

File: EIPS/eip-walletnft.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+The first use case that is devised is for value preservation. Digital assets, as NFTs, have value that has to be preserved in order to not be lost. If the asset has its own budget (in other ERC20 coins), could use it to autopreserve itself.
+
+## Specification
+The constructor should mint the unique token of the contract, and then the mint function should add a restriction to avoid further minting.
+
+Also, a `tokenTransfer` function should be added in order to allow the contract owner to transact with the ERC20 tokens owned by the contract/NFT itself.
+
+## Rationale
+The main motivation is to keep the contract compatible with current ERC721 platforms. That's why we maintain all the functions of the ERC721 standard even if they cannot be used (eg. mint) instead of just removing them.
```
Comment (Pandapip1) [inline EIPS/eip-walletnft.md:30]: Why not just use EIP173?

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
A few changes I'd suggest.

**Comment by victormunoz** [comment]
Thank you for your feedback!
The main difference with 173 is that our base is ERC721. This is an NFT with wallet, and the should be compatible with all ERC721 marketplaces platforms.
Current ERC721 implementation does not allow that the NFTs minted on it (tokenid's) to have its own balances in other ERC20, and that's why we propose this.

**Comment by Pandapip1** [comment]
But why would you want it to be an NFT? This spec really wants to be broken
up into two EIPs:
- An interface for a single NFT (like ownable but with metadata and
token-like qualities?)
- Have a smart contract wallet

Also, you'd want to have your smart contract wallet to be able to do way
more than just transfer ERC20 tokens. You'd probably want to make a
function that would allow the contract to perform many arbitrary calls in a
batch.

File: EIPS/eip-walletnft.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
+---
+eip: <to be assigned>
```
Comment (MicahZoltu) [inline EIPS/eip-walletnft.md:2]:
```suggestion
eip: 4944
```

**Review by victormunoz — COMMENTED** [review]
Reviewed and commited all the suggestions

File: EIPS/eip-4944.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+## Rationale
+The main motivation is to keep the contract compatible with current ERC721 platforms. That's why we maintain all the functions of the ERC721 standard even if they cannot be used (eg. mint) instead of just removing them.
+
+## Backwards Compatibility
+There are no backwards compatibility issues.
+
+## Reference Implementation
+Add the variable `_minted` in the contract:
+
+``` solidity
```
Comment (SamWilsn) [inline EIPS/eip-4944.md:39]: Your code blocks are all messed up.

File: EIPS/eip-4944.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+The first use case that is devised is for value preservation. Digital assets, as NFTs, have value that has to be preserved in order to not be lost. If the asset has its own budget (in other ERC20 coins), could use it to autopreserve itself.
+
+## Specification
+The constructor should mint the unique token of the contract, and then the mint function should add a restriction to avoid further minting.
+
+Also, a `tokenTransfer` function should be added in order to allow the contract owner to transact with the ERC20 tokens owned by the contract/NFT itself.
+
+## Rationale
+The main motivation is to keep the contract compatible with current ERC721 platforms. That's why we maintain all the functions of the ERC721 standard even if they cannot be used (eg. mint) instead of just removing them.
```
Comment (SamWilsn) [inline EIPS/eip-4944.md:31]: `mint` isn't part of ERC-721.

File: EIPS/eip-4944.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+This change would enable automatically all the capabilities of composable tokens ERC-998 (own other ERC721 or ERC20) natively without adding any extra code, just forbidding to mint more than one token per deployed contract.
+Then the NFT minted with this contract could operate with his "budget" (the ERC20 he owned) and also trade with the other NFTs he could own. Just like an autonomous agent, that could decide what to do with his properties (sell his NFTs, buy other NFTs, etc).
+
+The first use case that is devised is for value preservation. Digital assets, as NFTs, have value that has to be preserved in order to not be lost. If the asset has its own budget (in other ERC20 coins), could use it to autopreserve itself.
+
+## Specification
+The constructor should mint the unique token of the contract, and then the mint function should add a restriction to avoid further minting.
+
+Also, a `tokenTransfer` function should be added in order to allow the contract owner to transact with the ERC20 tokens owned by the contract/NFT itself.
```
Comment (SamWilsn) [inline EIPS/eip-4944.md:28]: Should probably expand on what this function does and how it works.

File: EIPS/eip-4944.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+## Motivation
+If the ERC721 was modified to mint only 1 token (per contract), then the contract address could be identified uniquely with that minted token (instead of the tuple contract address + token id, as ERC721 requires).
+This change would enable automatically all the capabilities of composable tokens ERC-998 (own other ERC721 or ERC20) natively without adding any extra code, just forbidding to mint more than one token per deployed contract.
+Then the NFT minted with this contract could operate with his "budget" (the ERC20 he owned) and also trade with the other NFTs he could own. Just like an autonomous agent, that could decide what to do with his properties (sell his NFTs, buy other NFTs, etc).
+
+The first use case that is devised is for value preservation. Digital assets, as NFTs, have value that has to be preserved in order to not be lost. If the asset has its own budget (in other ERC20 coins), could use it to autopreserve itself.
+
+## Specification
+The constructor should mint the unique token of the contract, and then the mint function should add a restriction to avoid further minting.
```
Comment (SamWilsn) [inline EIPS/eip-4944.md:26]: What is the `tokenId` of the minted token? Does it matter?

File: EIPS/eip-4944.md [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+## Abstract
+The following describes standard functions for an ERC721 compatible contract that can only only one minted token.
+This allows an NFT to be associated uniquely with a single contract address.
+
+## Motivation
+If the ERC721 was modified to mint only 1 token (per contract), then the contract address could be identified uniquely with that minted token (instead of the tuple contract address + token id, as ERC721 requires).
+This change would enable automatically all the capabilities of composable tokens ERC-998 (own other ERC721 or ERC20) natively without adding any extra code, just forbidding to mint more than one token per deployed contract.
+Then the NFT minted with this contract could operate with his "budget" (the ERC20 he owned) and also trade with the other NFTs he could own. Just like an autonomous agent, that could decide what to do with his properties (sell his NFTs, buy other NFTs, etc).
+
+The first use case that is devised is for value preservation. Digital assets, as NFTs, have value that has to be preserved in order to not be lost. If the asset has its own budget (in other ERC20 coins), could use it to autopreserve itself.
```
[the remark this answers is not in the document]
Comment (victormunoz) [reply EIPS/eip-4944.md:23]: The token does not disappear when it runs out of ERC-20 tokens. It simply cannot spend any more (in this use case), and eventually will provably become obsolete for sure

File: EIPS/eip-4944.md (hunk 12 above)
Comment (victormunoz) [reply EIPS/eip-4944.md:39]: Fixed

File: EIPS/eip-4944.md (hunk 13 above)
Comment (victormunoz) [reply EIPS/eip-4944.md]: Actually yes, you are right. I have deleted the sentence

File: EIPS/eip-4944.md (hunk 14 above)
Comment (victormunoz) [reply EIPS/eip-4944.md]: Expanded

File: EIPS/eip-4944.md (hunk 15 above)
Comment (victormunoz) [reply EIPS/eip-4944.md:26]: It is just there for compatibility with current ERC721 platforms.
The value should be 0 as suggested in the implementation section.

**Review by SamWilsn — APPROVED** [review]

**Comment by Pandapip1** [comment]
@SamWilsn fyi - this didn't merge.

**Comment by victormunoz** [comment]
This PR has turned to stagnant. Is there anything wrong with it? What can we do for it to be accepted?

**Comment by Pandapip1** [comment]
> This PR has turned to stagnant. Is there anything wrong with it? What can we do for it to be accepted?

Submit a PR moving it back to Draft. EIPs that have not been updated for 6+ months are automatically moved to stagnant.

**Comment by victormunoz** [comment]
How can I move it back to draft? Do you mean by creating a new PR? Or changing something in this one?

**Comment by Pandapip1** [comment]
Submit a new PR.

**Pull request merged**